File: backend/app/schemas/molecule.py
"""
Pydantic schemas for molecule-related requests and responses.
"""

from typing import Optional

from pydantic import BaseModel, Field

from app.core.config import settings


# ---------------------------------------------------------------------------
# Request
# ---------------------------------------------------------------------------
class SmilesRequest(BaseModel):
    smiles: str = Field(..., min_length=1, max_length=settings.SMILES_MAX_LENGTH)


# ---------------------------------------------------------------------------
# Response sub-models
# ---------------------------------------------------------------------------
class Physicochemical(BaseModel):
    formula: str
    mw: float
    heavy_atoms: int
    aromatic_heavy_atoms: int
    fraction_csp3: float
    rotatable_bonds: int
    hba: int
    hbd: int
    molar_refractivity: float
    tpsa: float


class LipophilicityData(BaseModel):
    crippen_logp: float


class SolubilityData(BaseModel):
    esol_logs: float
    esol_mg_ml: float
    esol_mol_l: float
    esol_class: str


class RuleResult(BaseModel):
    passes: bool
    violations: int
    details: list[str]


class DruglikenessData(BaseModel):
    lipinski: RuleResult
    ghose: RuleResult
    veber: RuleResult
    egan: RuleResult
    muegge: RuleResult
    bioavailability_score: float


class MedChemData(BaseModel):
    pains_alerts: int
    pains_descriptions: list[str]
    sa_score: Optional[float]
    leadlikeness: RuleResult


class RadarData(BaseModel):
    lipo: float
    size: float
    polar: float
    insolu: float
    insatu: float
    flex: float


# ---------------------------------------------------------------------------
# Top-level response
# ---------------------------------------------------------------------------
class MoleculeResponse(BaseModel):
    svg: str
    smiles: str
    physicochemical: Physicochemical
    lipophilicity: LipophilicityData
    solubility: SolubilityData
    druglikeness: DruglikenessData
    medicinal_chemistry: MedChemData
    radar: RadarData

File: frontend/types/molecule.ts
export interface Physicochemical {
  formula: string;
  mw: number;
  heavy_atoms: number;
  aromatic_heavy_atoms: number;
  fraction_csp3: number;
  rotatable_bonds: number;
  hba: number;
  hbd: number;
  molar_refractivity: number;
  tpsa: number;
}

export interface RuleResult {
  passes: boolean;
  violations: number;
  details: string[];
}

export interface RadarValues {
  lipo: number;
  size: number;
  polar: number;
  insolu: number;
  insatu: number;
  flex: number;
}

export interface MoleculeData {
  svg: string;
  smiles: string;
  physicochemical: Physicochemical;
  lipophilicity: { crippen_logp: number };
  solubility: {
    esol_logs: number;
    esol_mg_ml: number;
    esol_mol_l: number;
    esol_class: string;
  };
  druglikeness: {
    lipinski: RuleResult;
    ghose: RuleResult;
    veber: RuleResult;
    egan: RuleResult;
    muegge: RuleResult;
    bioavailability_score: number;
  };
  medicinal_chemistry: {
    pains_alerts: number;
    pains_descriptions: string[];
    sa_score: number | null;
    leadlikeness: RuleResult;
  };
  radar: RadarValues;
}
